File: backend/internal/model/rent.go
package model

import "time"

type PaymentScheme string

const (
	PaymentSchemeLumpSum     PaymentScheme = "LUMP_SUM"
	PaymentSchemeInstallment PaymentScheme = "INSTALLMENT"
)

type ContractTimeStatus string

const (
	ContractTimeActive    ContractTimeStatus = "ACTIVE"
	ContractTimeExpiring  ContractTimeStatus = "EXPIRING"
	ContractTimeEnded     ContractTimeStatus = "ENDED"
)

type ContractPaymentStatus string

const (
	ContractPaymentUnpaid  ContractPaymentStatus = "UNPAID"
	ContractPaymentPartial ContractPaymentStatus = "PARTIAL"
	ContractPaymentPaid    ContractPaymentStatus = "PAID"
)

type PeriodicContract struct {
	ID                 string                `json:"id"`
	WorkspaceID        string                `json:"workspaceId"`
	BusinessUnitID     string                `json:"businessUnitId"`
	BusinessUnitName   string                `json:"businessUnitName,omitempty"`
	StartDate          string                `json:"startDate"`
	EndDate            string                `json:"endDate"`
	DurationMonths     int                   `json:"durationMonths"`
	TotalAmount        float64               `json:"totalAmount"`
	PaymentScheme      PaymentScheme         `json:"paymentScheme"`
	MonthlyEquivalent  float64               `json:"monthlyEquivalent"`
	Notes              *string               `json:"notes,omitempty"`
	TimeStatus         ContractTimeStatus    `json:"timeStatus"`
	PaymentStatus      ContractPaymentStatus `json:"paymentStatus"`
	PaidAmount         float64               `json:"paidAmount"`
	RemainingAmount    float64               `json:"remainingAmount"`
	Schedules          []PaymentSchedule     `json:"schedules,omitempty"`
	CreatedAt          time.Time             `json:"createdAt"`
	UpdatedAt          time.Time             `json:"updatedAt"`
}

type PaymentSchedule struct {
	ID         string     `json:"id"`
	ContractID string     `json:"contractId"`
	DueDate    string     `json:"dueDate"`
	Amount     float64    `json:"amount"`
	IsPaid     bool       `json:"isPaid"`
	PaidAt     *time.Time `json:"paidAt,omitempty"`
}

type ContractPayment struct {
	ID            string    `json:"id"`
	ScheduleID    string    `json:"scheduleId"`
	TransactionID string    `json:"transactionId"`
	Amount        float64   `json:"amount"`
	PaymentDate   string    `json:"paymentDate"`
	Notes         *string   `json:"notes,omitempty"`
	CreatedAt     time.Time `json:"createdAt"`
}

File: frontend/src/api/rent.ts
import { api } from '@/api/client'

export type PaymentScheme = 'LUMP_SUM' | 'INSTALLMENT'
export type ContractTimeStatus = 'ACTIVE' | 'EXPIRING' | 'ENDED'
export type ContractPaymentStatus = 'UNPAID' | 'PARTIAL' | 'PAID'

export type PaymentSchedule = {
  id: string
  contractId: string
  dueDate: string
  amount: number
  isPaid: boolean
  paidAt?: string
}

export type PeriodicContract = {
  id: string
  businessUnitId: string
  businessUnitName?: string
  startDate: string
  endDate: string
  durationMonths: number
  totalAmount: number
  paymentScheme: PaymentScheme
  monthlyEquivalent: number
  notes?: string
  timeStatus: ContractTimeStatus
  paymentStatus: ContractPaymentStatus
  paidAmount: number
  remainingAmount: number
  schedules?: PaymentSchedule[]
}

export type CreateRentContractInput = {
  businessUnitId: string
  startDate: string
  durationMonths: number
  totalAmount: number
  paymentScheme: PaymentScheme
  notes?: string
}

export function listRentContracts(params?: { businessUnitId?: string }) {
  const query = new URLSearchParams()
  if (params?.businessUnitId) query.set('businessUnitId', params.businessUnitId)
  const suffix = query.toString() ? `?${query}` : ''
  return api.get<PeriodicContract[]>(`/rent-contracts${suffix}`)
}

export function getRentContract(id: string) {
  return api.get<PeriodicContract>(`/rent-contracts/${id}`)
}

export function createRentContract(body: CreateRentContractInput) {
  return api.post<{ contract: PeriodicContract; schedules: PaymentSchedule[] }>('/rent-contracts', body)
}

export function payRentSchedule(
  scheduleId: string,
  body: { paymentDate: string; cashAccountId: string; notes?: string },
) {
  return api.post<{ payment: unknown; transaction: { id: string } }>(
    `/rent-contracts/schedules/${scheduleId}/pay`,
    body,
  )
}

export const paymentSchemeLabel: Record<PaymentScheme, string> = {
  LUMP_SUM: 'Sekaligus',
  INSTALLMENT: 'Cicilan bulanan',
}

export const timeStatusLabel: Record<ContractTimeStatus, string> = {
  ACTIVE: 'Aktif',
  EXPIRING: 'Hampir berakhir',
  ENDED: 'Berakhir',
}

export const paymentStatusLabel: Record<ContractPaymentStatus, string> = {
  UNPAID: 'Belum bayar',
  PARTIAL: 'Sebagian',
  PAID: 'Lunas',
}
